include/realtime_tools/realtime_box.h

```diff
@@ -26,50 +26,251 @@
 // ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 // POSSIBILITY OF SUCH DAMAGE.
 
-// Author: Stuart Glaser
+// Author: Lennart Nachtigall
 
-#ifndef REALTIME_TOOLS__REALTIME_BOX_H__
-#define REALTIME_TOOLS__REALTIME_BOX_H__
+#ifndef REALTIME_TOOLS__REALTIME_BOX_H_
+#define REALTIME_TOOLS__REALTIME_BOX_H_
 
+#include <functional>
+#include <initializer_list>
 #include <mutex>
-#include <string>
+#include <optional>
+#include <utility>
+
+#include <rcpputils/pointer_traits.hpp>
 
 namespace realtime_tools
 {
+
+template <typename T>
+constexpr auto is_ptr_or_smart_ptr = rcpputils::is_pointer<T>::value;
+
 /*!
+    A Box that ensures thread safe access to the boxed contents.
+    Access is best effort. If it can not lock it will return.
 
-  Strongly suggested that you use an std::shared_ptr in this box to
-  guarantee realtime safety.
+    NOTE about pointers:
+    You can use pointers with this box but the access will be different.
+    Only use the get/set methods that take function pointer for accessing the internal value.
+*/
 
- */
-template <class T>
+// Provide a specialisation for non pointer types
+// NOTE: When migrating to a safe access only version just remove the specialisation for pointer
+// and let this be the only version!
+template <class T, typename mutex_type = std::mutex>
 class RealtimeBox
 {
+  static_assert(
+    std::is_same_v<mutex_type, std::mutex> || std::is_same_v<mutex_type, std::recursive_mutex>);
```

Comment:
@firesurfer do we have a need for recursive mutex here?, may be it is related to the `get_mutex` methods you implemented.

If so, can we rename this class to `RealTimeBoxBase` and with `using` keyword we create instances of different mutexes. The cool thing with that is when we have the priority inheritance mutexes, we can simply replace there and everyone using that code will have it propagated. 

What do you think?

Reply:
Well this class already allows to template the type of mutex. I thought it might be a good idea to limit which mutexes are allowed. But we can remove this static_assert

Reply:
> If so, can we rename this class to RealTimeBoxBase and with `using` keyword we create instances of different mutexes. 

What do you think about this approach? And have things with using? The default RealTimeBox instance will use std::mutex as here

Reply:
I just wanted to understand the why :)

I can add them, as well as a using for the RealTimeBoxBestEffort with a deprecation notice.

Reply:
If you are convinced, let's do it. If not, not a problem 😉

Reply:
So I added them. Please take a look if you agree with the names....naming things is always hard :D 

And @fmauch should now get a nice deprecation warning for the RealTimeBoxBestEffort